---
title: "VariantPlots"
author: "Nae-Chyun"
date: "3/9/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggpubr)
library(reshape2)
library(RColorBrewer)
library(hash)
library(ggridges)
library(scales)
library(ggrepel)
source("helper.R")
```

```{r global}
save <- FALSE

cbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#999999")

# make precision-recall curve
pr <- T

# filenames <- c("chm13_to_grch38", "grch38", "grch38_masked", "chm13_to_hg19", "hg19")
# printednames <- c("CHM13-to-GRCh38", "GRCh38", "GRCh38-masked", "CHM13-to-GRCh37", "GRCh37")
filenames <- c("grch38", "grch38_masked", "chm13_to_grch38", "hg19", "chm13_to_hg19", "grch37", "chm13_to_grch37")
printednames <- c("GRCh38", "GRCh38-masked", "CHM13-to-GRCh38", "GRCh37", "CHM13-to-GRCh37", "GRCh37", "CHM13-to-GRCh37")
levelnames <- c("GRCh38", "GRCh38-masked", "CHM13-to-GRCh38", "GRCh37", "CHM13-to-GRCh37")
map_caller_name <- hash()
for (i in seq(1, length(filenames))){
  map_caller_name[[filenames[i]]] <- printednames[i]}

grc_systems <- c("GRCh38", "GRCh38", "GRCh38", "GRCh37", "GRCh37", "GRCh37", "GRCh37", "GRCh37")
map_grc_system <- hash()
for (i in seq(1, length(printednames))){
  map_grc_system[[printednames[i]]] <- grc_systems[i]}

aln_short <- c("bt2", "bwa")
aln_names <- c("Bowtie 2", "BWA-MEM")
map_aln_name <- hash()
for (i in seq(1, length(aln_short))){
  map_aln_name[[aln_short[i]]] <- aln_names[i]}

metric_orig <- c("METRIC.Recall", "METRIC.Precision", "METRIC.F1_Score", "TRUTH.TP", "TRUTH.FN", "QUERY.FP")
metric_names <- c("Recall", "Precision", "F1", "TP", "FN", "FP")
map_metrics <- hash()
for (i in seq(1, length(metric_orig))){
  map_metrics[[metric_orig[i]]] <- metric_names[i]}

metric_sv_orig <- c("recall", "precision", "f1", "TRUTH.TP", "TRUTH.FN", "QUERY.FP")
metric_sv_names <- c("Recall", "Precision", "F1", "TP", "FN", "FP")
map_metrics_sv <- hash()
for (i in seq(1, length(metric_sv_orig))){
  map_metrics[[metric_sv_orig[i]]] <- metric_sv_names[i]}
```

```{r small_variant_wgs_dv, message=FALSE}
df.hg001.bwa.dv <- do.call(rbind, lapply(c("chm13v2/hg001/grch38.summary.csv", "chm13v2/hg001/chm13_to_grch38.summary.csv", "chm13v2/hg001/hg19.summary.csv", "chm13v2/hg001/chm13_to_hg19.summary.csv"), read_csv, map=map_caller_name))
df.hg002.bwa.dv <- do.call(rbind, lapply(c("chm13v2/hg002/grch38.summary.csv", "chm13v2/hg002/chm13_to_grch38.summary.csv", "chm13v2/hg002/hg19.summary.csv", "chm13v2/hg002/chm13_to_hg19.summary.csv"), read_csv, map=map_caller_name))
df.hg005.bwa.dv <- do.call(rbind, lapply(c("chm13v2/hg005/grch38.summary.csv", "chm13v2/hg005/chm13_to_grch38.summary.csv", "chm13v2/hg005/hg19.summary.csv", "chm13v2/hg005/chm13_to_hg19.summary.csv"), read_csv, map=map_caller_name))

df.hg001.bwa.dv$Sample <- "HG001"
df.hg002.bwa.dv$Sample <- "HG002"
df.hg005.bwa.dv$Sample <- "HG005"
df.bwa.dv <- rbind(df.hg002.bwa.dv, df.hg001.bwa.dv, df.hg005.bwa.dv) %>% filter(Filter == "PASS")
df.bwa.dv$Sample <- factor(df.bwa.dv$Sample, levels=c("HG001", "HG002", "HG005"))
df.bwa.dv$name <- factor(df.bwa.dv$name, levels=levelnames)
df.bwa.dv$name_rev <- factor(df.bwa.dv$name, levels=rev(levelnames))
df.bwa.dv$GrcSystem <- ifelse(df.bwa.dv$name %in% c("GRCh38", "CHM13-to-GRCh38"), "GRCh38", "GRCh37")

# df.bwa.dv.melt <- melt(df.bwa.dv %>% select(name, Sample, Type, METRIC.Recall, METRIC.Precision, METRIC.F1_Score))
df.bwa.dv.melt <- melt(df.bwa.dv %>% select(name, Sample, Type, METRIC.F1_Score))
df.bwa.dv.melt$variable <- factor(hash::values(map_metrics, keys=df.bwa.dv.melt$variable), levels = metric_names)
df.bwa.dv.melt$GrcSystem <- factor(hash::values(map_grc_system, keys=df.bwa.dv.melt$name))
fig.bwa.dv.f1 <- ggplot(df.bwa.dv.melt, aes(x=variable, y=value, fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) +
    scale_fill_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
    facet_grid(GrcSystem~Type+Sample) +
    # facet_grid(Type~.) +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + coord_cartesian(ylim=c(0.99,1)) +
    theme(axis.title.x=element_text(vjust=0, hjust=0.45), legend.position = "bottom", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.bwa.dv.f1

df.bwa.dv %>% filter(Filter == "PASS") %>% mutate(Err = TRUTH.FN + QUERY.FP) %>% select(name, Type, metric_orig, Err)

df.bwa.dv <- df.bwa.dv %>% mutate(GrcSystem = ifelse(name %in% c("GRCh38", "CHM13-to-GRCh38"), "GRCh38", "GRCh37"), pair = (name %in% c("GRCh38", "CHM13-to-GRCh38") * 1) + (Type == "SNP") * 10)

fig.bwa.dv.scatter <- ggplot(df.bwa.dv, aes(y=METRIC.Recall, x=METRIC.Precision, color=name, shape=Type)) + 
  geom_path(aes(group = pair), color="#999999", arrow = arrow(length=unit(0.055, "inches"))) +
  geom_point() + 
  facet_grid(~GrcSystem+Sample, labeller = label_wrap_gen(multi_line=FALSE)) +
  scale_color_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
  coord_cartesian(ylim=c(0.986,1), xlim=c(0.986,1)) +
  ylab("Recall") + xlab("Precision") + labs(color="Method") + theme_bw() +
  theme(axis.title.x=element_text(vjust=0, hjust=0.45),
        axis.text.x = element_text(angle = 90, vjust = 1, hjust=0.8),
        legend.position = "bottom", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.bwa.dv.scatter

if (save){
  ggsave("figures/0309/small_variants-bwa_dv-f1.pdf", fig.bwa.dv.f1
         + theme(legend.position = "none")
         , width=4, height=4)
  ggsave("figures/0309/small_variants-bwa_dv-scatter.pdf", fig.bwa.dv.scatter
         + theme(legend.position = "none")
         , width=8, height=4)
  # write.csv(
  #   df.bwa.dv %>% filter(Filter=="PASS") %>% select(name, Type, Sample, TRUTH.TP, TRUTH.FN, QUERY.FP, METRIC.Recall, METRIC.Precision, METRIC.F1_Score),
  #   "csvs/small_variants-bwa_dv.csv", row.names = FALSE)

}
```

```{r small_variant_wgs_gatk, message=FALSE}
df.hg001.bwa.gatk <- do.call(rbind, lapply(c("chm13v2-gatk/hg001/grch38.summary.csv", "chm13v2-gatk/hg001/chm13_to_grch38.summary.csv", "chm13v2-gatk/hg001/hg19.summary.csv", "chm13v2-gatk/hg001/chm13_to_hg19.summary.csv"), read_csv, map=map_caller_name))
df.hg002.bwa.gatk <- do.call(rbind, lapply(c("chm13v2-gatk/hg002/grch38.summary.csv", "chm13v2-gatk/hg002/chm13_to_grch38.summary.csv", "chm13v2-gatk/hg002/hg19.summary.csv", "chm13v2-gatk/hg002/chm13_to_hg19.summary.csv"), read_csv, map=map_caller_name))
df.hg005.bwa.gatk <- do.call(rbind, lapply(c("chm13v2-gatk/hg005/grch38.summary.csv", "chm13v2-gatk/hg005/chm13_to_grch38.summary.csv", "chm13v2-gatk/hg005/hg19.summary.csv", "chm13v2-gatk/hg005/chm13_to_hg19.summary.csv"), read_csv, map=map_caller_name))


df.hg001.bwa.gatk$Sample <- "HG001"
df.hg002.bwa.gatk$Sample <- "HG002"
df.hg005.bwa.gatk$Sample <- "HG005"
df.bwa.gatk <- rbind(df.hg002.bwa.gatk, df.hg001.bwa.gatk, df.hg005.bwa.gatk) %>% filter(Filter == "PASS")
df.bwa.gatk$Sample <- factor(df.bwa.gatk$Sample, levels=c("HG001", "HG002", "HG005"))
df.bwa.gatk$name <- factor(df.bwa.gatk$name, levels=levelnames)
df.bwa.gatk$name_rev <- factor(df.bwa.gatk$name, levels=rev(levelnames))
df.bwa.gatk$GrcSystem <- ifelse(df.bwa.gatk$name %in% c("GRCh38", "CHM13-to-GRCh38"), "GRCh38", "GRCh37")

# df.bwa.gatk.melt <- melt(df.bwa.gatk %>% select(name, Sample, Type, METRIC.Recall, METRIC.Precision, METRIC.F1_Score))
df.bwa.gatk.melt <- melt(df.bwa.gatk %>% select(name, Sample, Type, METRIC.F1_Score))
df.bwa.gatk.melt$variable <- factor(hash::values(map_metrics, keys=df.bwa.gatk.melt$variable), levels = metric_names)
df.bwa.gatk.melt$GrcSystem <- factor(hash::values(map_grc_system, keys=df.bwa.gatk.melt$name))
fig.bwa.gatk.f1 <- ggplot(df.bwa.gatk.melt, aes(x=variable, y=value, fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) +
    scale_fill_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
    facet_grid(GrcSystem~Type+Sample) +
    # facet_grid(Type~.) +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + coord_cartesian(ylim=c(0.98,1)) +
    theme(axis.title.x=element_text(vjust=0, hjust=0.45), legend.position = "bottom", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.bwa.gatk.f1

df.bwa.gatk %>% filter(Filter == "PASS") %>% mutate(Err = TRUTH.FN + QUERY.FP) %>% select(name, Type, metric_orig, Err)

df.bwa.gatk <- df.bwa.gatk %>% mutate(GrcSystem = ifelse(name %in% c("GRCh38", "CHM13-to-GRCh38"), "GRCh38", "GRCh37"), pair = (name %in% c("GRCh38", "CHM13-to-GRCh38") * 1) + (Type == "SNP") * 10)

fig.bwa.gatk.scatter <- ggplot(df.bwa.gatk, aes(y=METRIC.Recall, x=METRIC.Precision, color=name, shape=Type)) + 
  geom_path(aes(group = pair), color="#999999", arrow = arrow(length=unit(0.055, "inches"))) +
  geom_point() + 
  facet_grid(~GrcSystem+Sample, labeller = label_wrap_gen(multi_line=FALSE)) +
  scale_color_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
  coord_cartesian(ylim=c(0.975,1), xlim=c(0.975,1)) +
  ylab("Recall") + xlab("Precision") + labs(color="Method") + theme_bw() +
  theme(axis.title.x=element_text(vjust=0, hjust=0.45),
        axis.text.x = element_text(angle = 90, vjust = 1, hjust=0.8),
        legend.position = "bottom", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.bwa.gatk.scatter

if (save){
  ggsave("figures/0309/small_variants-bwa_gatk-f1.pdf", fig.bwa.gatk.f1
         + theme(legend.position = "none")
         , width=4, height=4)
  ggsave("figures/0309/small_variants-bwa_gatk-scatter.pdf", fig.bwa.gatk.scatter
         + theme(legend.position = "none")
         , width=8, height=4)
  # write.csv(
  #   df.bwa.dv %>% filter(Filter=="PASS") %>% select(name, Type, Sample, TRUTH.TP, TRUTH.FN, QUERY.FP, METRIC.Recall, METRIC.Precision, METRIC.F1_Score),
  #   "csvs/small_variants-bwa_dv.csv", row.names = FALSE)

}
```


```{r small_variant_wgs_dv_cmrg_hg002}
df_hg002_bwa_dv_cmrg <- do.call(rbind, lapply(c("chm13v2/hg002/cmrg/grch38.summary.csv", "chm13v2/hg002/cmrg/chm13_to_grch38.summary.csv", "chm13v2/hg002/cmrg/grch37.summary.csv", "chm13v2/hg002/cmrg/chm13_to_grch37.summary.csv"), read_csv, map=map_caller_name))
df_hg002_bwa_dv_cmrg$name = factor(df_hg002_bwa_dv_cmrg$name, levels=levelnames)

df_hg002_bwa_dv_cmrg_melt <- melt(df_hg002_bwa_dv_cmrg %>% select(name, Type, METRIC.Recall, METRIC.Precision, METRIC.F1_Score))#metric_orig))
df_hg002_bwa_dv_cmrg_melt$variable <- factor(hash::values(map_metrics, keys=df_hg002_bwa_dv_cmrg_melt$variable), levels = metric_names)
df_hg002_bwa_dv_cmrg_melt$GrcSystem <- factor(hash::values(map_grc_system, keys=df_hg002_bwa_dv_cmrg_melt$name))
fig_hg002_bwa_dv_cmrg <- ggplot(df_hg002_bwa_dv_cmrg_melt, aes(x=variable, y=value, fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) +
    scale_fill_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
    facet_grid(GrcSystem~Type) +
    # facet_grid(Type~.) +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + coord_cartesian(ylim=c(0.9,1)) +
    theme(axis.title.x=element_text(vjust=0, hjust=0.45), legend.position = "none", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig_hg002_bwa_dv_cmrg 

df_hg002_bwa_dv_cmrg %>% filter(Filter == "PASS") %>% mutate(Err = TRUTH.FN + QUERY.FP) %>% select(name, Type, metric_orig, Err)

if (save){
  ggsave("figures/0309/small_variants-bwa_dv-cmrg.pdf", fig_hg002_bwa_dv_cmrg, width=5, height=5)
}
```

```{r small_variant_wgs_gatk_cmrg_hg002}
df.hg002.bwa.gatk.cmrg <- do.call(rbind, lapply(c("chm13v2-gatk/hg002/cmrg/grch38.summary.csv", "chm13v2-gatk/hg002/cmrg/chm13_to_grch38.summary.csv", "chm13v2-gatk/hg002/cmrg/hg19.summary.csv", "chm13v2-gatk/hg002/cmrg/chm13_to_hg19.summary.csv"), read_csv, map=map_caller_name)) %>% filter(Filter == "PASS")
df.hg002.bwa.gatk.cmrg$name = factor(df.hg002.bwa.gatk.cmrg$name, levels=levelnames)

df.hg002.bwa.gatk.cmrg.melt <- melt(df.hg002.bwa.gatk.cmrg %>% select(name, Type, METRIC.Recall, METRIC.Precision, METRIC.F1_Score))#metric_orig))
df.hg002.bwa.gatk.cmrg.melt$variable <- factor(hash::values(map_metrics, keys=df.hg002.bwa.gatk.cmrg.melt$variable), levels = metric_names)
df.hg002.bwa.gatk.cmrg.melt$GrcSystem <- factor(hash::values(map_grc_system, keys=df.hg002.bwa.gatk.cmrg.melt$name))
fig.hg002.bwa.gatk.cmrg <- ggplot(df.hg002.bwa.gatk.cmrg.melt, aes(x=variable, y=value, fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) +
    scale_fill_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
    facet_grid(GrcSystem~Type) +
    # facet_grid(Type~.) +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + coord_cartesian(ylim=c(0.9,1)) +
    theme(axis.title.x=element_text(vjust=0, hjust=0.45), legend.position = "none", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.hg002.bwa.gatk.cmrg 

df.hg002.bwa.gatk.cmrg %>% filter(Filter == "PASS") %>% mutate(Err = TRUTH.FN + QUERY.FP) %>% select(name, Type, metric_orig, Err)

if (save){
  ggsave("figures/0309/small_variants-bwa_gatk-cmrg.pdf", fig.hg002.bwa.gatk.cmrg, width=5, height=5)
}
```



```{r stratification_settings}
strat_subsets <- c('*', 'alldifficultregions', 'alllowmapandsegdupregions', 'lowmappabilityall', 'segdups', 'allOtherDifficultregions', 
                   'AllHomopolymers_gt6bp_imperfectgt10bp_slop5', 'AllTandemRepeats_201to10000bp_slop5', 'AllTandemRepeats_51to200bp_slop5', 'AllTandemRepeats_gt10000bp_slop5', 'AllTandemRepeats_gt100bp_slop5', 'AllTandemRepeats_lt51bp_slop5', 'AllTandemRepeatsandHomopolymers_slop5', 'BadPromoters', 'CMRGv1.00_duplicationinKMT2C', 'CMRGv1.00_falselyduplicatedgenes', 'KIR', 'L1H_gt500', 'LD_discordant_haplotypes_slop5bp', 'MHC', 'SimpleRepeat_diTR_11to50_slop5', 'SimpleRepeat_diTR_51to200_slop5', 'SimpleRepeat_diTR_gt200_slop5', 'SimpleRepeat_homopolymer_4to6_slop5', 'SimpleRepeat_homopolymer_7to11_slop5', 'SimpleRepeat_homopolymer_gt11_slop5', 'SimpleRepeat_homopolymer_gt20_slop5', 'SimpleRepeat_imperfecthomopolgt10_slop5', 'SimpleRepeat_imperfecthomopolgt20_slop5', 'SimpleRepeat_quadTR_20to50_slop5', 'SimpleRepeat_quadTR_51to200_slop5', 'SimpleRepeat_quadTR_gt200_slop5', 'SimpleRepeat_triTR_15to50_slop5', 'SimpleRepeat_triTR_51to200_slop5', 'SimpleRepeat_triTR_gt200_slop5', 'VDJ', 'ancestry_AFR', 'ancestry_AMR', 'ancestry_EAS', 'ancestry_EUR', 'ancestry_Neanderthal', 'ancestry_SAS', 'chainSelf', 'chainSelf_gt10kb', 'collapsed_duplication_FP_regions', 'contigs_lt500kb', 'false_duplications_correct_copy', 'false_duplications_incorrect_copy', 'gaps_slop15kb', 'gc15_slop50', 'gc15to20_slop50', 'gc20to25_slop50', 'gc25to30_slop50', 'gc30to55_slop50', 'gc55to60_slop50', 'gc60to65_slop50', 'gc65to70_slop50', 'gc70to75_slop50', 'gc75to80_slop50', 'gc80to85_slop50', 'gc85_slop50', 'gclt25orgt65_slop50', 'gclt30orgt55_slop50', 'gnomAD_InbreedingCoeff_slop1bp_merge1000bp', 'gt5segdups_gt10kb_gt99percidentity', 'hg38_minimap2_asm20_N10_gt1contig_gt1kb', 'hg38_minimap2_asm20_N10_nocovgt1kb', 'hs37d5_decoy_alignments',  'missing_and_multiple_alignments_of_GRCh38', 'nonunique_l100_m2_e1', 'nonunique_l250_m0_e0', 'notinAllHomopolymers_gt6bp_imperfectgt10bp_slop5', 'notinAllTandemRepeatsandHomopolymers_slop5', 'notinalldifficultregions', 'notinalllowmapandsegdupregions', 'notinchainSelf', 'notinchainSelf_gt10kb', 'notinlowmappabilityall', 'notinrefseq_cds', 'notinsegdups', 'notinsegdups_gt10kb', 'population_CNV_FP_regions', 'refseq_cds', 'segdups_gt10kb')
strat_names <- c('All', 'AllDifficult', 'LowMap&SegDup', 'LowMap', 'SegDups', 'AllOtherDifficult', 
                 'Homo>6bp_imperfectgt10bp', 'Tandem201-10000bp', 'Tandem51-200bp', 'Tandem>10000bp', 'Tandem>100bp', 'Tandem<51bp', 'Tandem&Homo', 'BadPromoters', 'CMRGv1.00_duplicationinKMT2C', 'CMRGv1.00_falselyduplicatedgenes', 'KIR', 'L1H>500', 'LD_discordant_haplotypes', 'MHC', 'DiTR_11-50', 'DiTR_51-200', 'DiTR>200', 'Homo4-6bp', 'Homo7-11bp', 'Homo>11bp', 'Homo>20bp', 'ImperfectHomo>10bp', 'ImperfectHomo>20bp', 'QuadTR_20-50', 'QuadTR_51-200', 'QuadTR>200', 'TriTR_15-50', 'TriTR_51-200', 'TriTR>200', 'VDJ', 'AFR', 'AMR', 'EAS', 'EUR', 'Neanderthal', 'SAS', 'ChainSelf', 'ChainSelf>10kb', 'collapsed_duplication_FP', 'contigs<500kb', 'false_duplications_correct_copy', 'false_duplications_incorrect_copy', 'gaps_slop15kb', 'GC15', 'GC15-20', 'GC20-25', 'GC25-30', 'GC30-55', 'GC55-60', 'GC60-65', 'GC65-70', 'GC70-75', 'GC75-80', 'GC80-85', 'GC85', 'GC<25or>65', 'GC<30or>55', 'gnomAD_InbreedingCoeff_slop1bp_merge1000bp', '>5SegDups>10kb>99percidentity', 'hg38_minimap2_asm20_N10>1contig>1kb', 'hg38_minimap2_asm20_N10_nocov>1kb', 'hs37d5Decoy', 'missing_and_multiple_alignments_of_GRCh38', 'NonUnique100bp', 'NonUnique250bp', 'NotInHomo>6bp_imperfect>10bp', 'NotInTandem&Homo', 'NotInAllDifficult', 'NotInLowMap&SegDup', 'NotInChainSelf', 'NotInChainSelf>10kb', 'NotInLowMap', 'NotInRefSeqCds', 'NotInSegDups', 'NotInSegDups>10kb', 'population_CNV_FP_regions', 'RefSeqCds', 'SegDups>10kb')
map_strat_subsets <- hash()
for (i in seq(1, length(strat_subsets))){
  map_strat_subsets[[strat_subsets[i]]] <- strat_names[i]}

stratify_subsets <- 
# Mixed
# c("*", "alldifficultregions", "alllowmapandsegdupregions", "lowmappabilityall", "segdups")
  c("alldifficultregions", "lowmappabilityall", "segdups")
  # c("alldifficultregions", "lowmappabilityall", "segdups", "allOtherDifficultregions")
  # c("*", "alldifficultregions", "lowmappabilityall", "segdups", "allOtherDifficultregions")
  # c("*", "allOtherDifficultregions", "alldifficultregions", "alllowmapandsegdupregions", "segdups")
# FunctionalRegions
# c("refseq_cds", "notinrefseq_cds")
# OtherDifficult (false duplications)
# c("false_duplications_correct_copy", "false_duplications_incorrect_copy")
# c("AllHomopolymers_gt6bp_imperfectgt10bp_slop5", "AllTandemRepeats_gt100bp_slop5", "AllTandemRepeatsandHomopolymers_slop5")
# SegmentalDuplications (self chain)
  # c("chainSelf", "chainSelf_gt10kb", "notinchainSelf", "notinchainSelf_gt10kb")
# SegmentalDuplications (self chain)
  # c("notinsegdups_gt10kb", "notinsegdups", "segdups_gt10kb")
# c("MHC", "VDJ", "KIR")
# "contigs_lt500kb", "gaps_slop15kb", "nonunique_l250_m0_e0", 
# c("lowmappabilityall", "nonunique_l100_m2_e1", "nonunique_l250_m0_e0", "notinlowmappabilityall")
# GRC issues
# c("false_duplications_correct_copy", "false_duplications_incorrect_copy", "hs37d5_decoy_alignments")
# Invalid
# "gt5segdups_gt10kb_gt99percidentity", 

topdiff_n <- 10 #5

color_codes <- c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")

```

```{r small_variant_calling_stratified_hg002_compare}
df_hg002_bwa_dv_stratify_grc <- do.call(rbind, lapply(c("chm13v2/hg002/stratify/grch38.extended.csv", "chm13v2/hg002/stratify/grch37.extended.csv"), read_csv, map=map_caller_name)) %>% filter(Subtype == "*", Filter == "PASS", Subset.IS_CONF.Size > 10e6, Subset %in% strat_subsets)
df_hg002_bwa_dv_stratify_leviosam <- do.call(rbind, lapply(c("chm13v2/hg002/stratify/chm13_to_grch38.extended.csv", "chm13v2/hg002/stratify/chm13_to_grch37.extended.csv"), read_csv, map=map_caller_name)) %>% filter(Subtype == "*", Filter == "PASS", Subset.IS_CONF.Size > 10e6, Subset %in% strat_subsets)

# df_hg002_bwa_dv_stratify_grc <- do.call(rbind, lapply(c("chm13v2/hg002/stratify/grch38.extended.csv", "chm13v2/hg002/stratify/grch37.extended.csv"), read_csv, map=map_caller_name)) %>% filter(Subtype == "*", Filter == "PASS", Subset.IS_CONF.Size > 10e6, Subset %in% strat_subsets, Type == "INDEL")
# df_hg002_bwa_dv_stratify_leviosam <- do.call(rbind, lapply(c("chm13v2/hg002/stratify/chm13_to_grch38.extended.csv", "chm13v2/hg002/stratify/chm13_to_grch37.extended.csv"), read_csv, map=map_caller_name)) %>% filter(Subtype == "*", Filter == "PASS", Subset.IS_CONF.Size > 10e6, Subset %in% strat_subsets, Type == "INDEL")

df_hg002_bwa_dv_stratify_grc$Subset.Rename <- factor(hash::values(map_strat_subsets, keys=df_hg002_bwa_dv_stratify_grc$Subset), levels = strat_names)
df_hg002_bwa_dv_stratify_leviosam$Subset.Rename <- factor(hash::values(map_strat_subsets, keys=df_hg002_bwa_dv_stratify_leviosam$Subset), levels = strat_names)
# df_hg002_bwa_dv_stratify_leviosam <- df_hg002_bwa_dv_stratify_leviosam %>% filter(Subtype == "*", Filter == "PASS", Subset.IS_CONF.Size > 5e6)
df_hg002_bwa_dv_stratify_grc$ErrDiff <- df_hg002_bwa_dv_stratify_grc$TRUTH.FN + df_hg002_bwa_dv_stratify_grc$QUERY.FP - df_hg002_bwa_dv_stratify_leviosam$TRUTH.FN - df_hg002_bwa_dv_stratify_leviosam$QUERY.FP
df_hg002_bwa_dv_stratify_leviosam$ErrDiff <- df_hg002_bwa_dv_stratify_grc$TRUTH.FN + df_hg002_bwa_dv_stratify_grc$QUERY.FP - df_hg002_bwa_dv_stratify_leviosam$TRUTH.FN - df_hg002_bwa_dv_stratify_leviosam$QUERY.FP
df_hg002_bwa_dv_stratify_grc <- df_hg002_bwa_dv_stratify_grc %>% mutate(ConfSubsetSize_Mbp = Subset.IS_CONF.Size / 1000000)
df_hg002_bwa_dv_stratify_grc <- df_hg002_bwa_dv_stratify_grc %>% mutate(ErrDiff, ErrDiffDensity = ErrDiff / ConfSubsetSize_Mbp)
df_hg002_bwa_dv_stratify_leviosam <- df_hg002_bwa_dv_stratify_leviosam %>% mutate(ConfSubsetSize_Mbp = Subset.IS_CONF.Size / 1000000)
df_hg002_bwa_dv_stratify_leviosam <- df_hg002_bwa_dv_stratify_leviosam %>% mutate(ErrDiff, ErrDiffDensity = ErrDiff / ConfSubsetSize_Mbp)

df_hg002_bwa_dv_stratify <- rbind(df_hg002_bwa_dv_stratify_grc, df_hg002_bwa_dv_stratify_leviosam)


dat.hg002.small.var.stratify.grc.topdiff.melt.h37 <- melt(rbind(df_hg002_bwa_dv_stratify_grc %>% filter(name == "GRCh37", ErrDiffDensity > 1) %>% top_n(topdiff_n), df_hg002_bwa_dv_stratify_grc %>% filter(name == "GRCh37", ErrDiffDensity < -1) %>% top_n(-topdiff_n)) %>% select(Type, Subtype, Subset.Rename, ConfSubsetSize_Mbp, ErrDiffDensity, name))
dat.hg002.small.var.stratify.grc.topdiff.melt.h37$reorder <- rep(c(dat.hg002.small.var.stratify.grc.topdiff.melt.h37 %>% filter(variable == "ErrDiffDensity"))$value, 2)
dat.hg002.small.var.stratify.grc.topdiff.melt.h37$variable <- factor(
  dat.hg002.small.var.stratify.grc.topdiff.melt.h37$variable,
  levels = c("ErrDiffDensity", "ConfSubsetSize_Mbp"),
  labels = c("#Error Diff Per Mbp", "SubSetSize (Mbp)"))
fig_hg002_stratify_topdiff_h37 <- ggplot(dat.hg002.small.var.stratify.grc.topdiff.melt.h37, aes(x=value, y=reorder(Subset.Rename, reorder), fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) + scale_fill_manual(values = color_codes[3]) +
    facet_grid(~variable, scales = "free") +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + theme(legend.position = "none", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig_hg002_stratify_topdiff_h37

dat.hg002.small.var.stratify.grc.topdiff.melt.h38 <- melt(
  rbind(df_hg002_bwa_dv_stratify_grc %>% filter(name == "GRCh38", ErrDiffDensity > 1) %>% top_n(topdiff_n), 
        df_hg002_bwa_dv_stratify_grc %>% filter(name == "GRCh38", ErrDiffDensity < -1) %>% top_n(-topdiff_n)) %>% select(Type, Subtype, Subset.Rename, ConfSubsetSize_Mbp, ErrDiffDensity, name))
dat.hg002.small.var.stratify.grc.topdiff.melt.h38$reorder <- rep(c(dat.hg002.small.var.stratify.grc.topdiff.melt.h38 %>% filter(variable == "ErrDiffDensity"))$value, 2)
dat.hg002.small.var.stratify.grc.topdiff.melt.h38$variable <- factor(
  dat.hg002.small.var.stratify.grc.topdiff.melt.h38$variable,
  levels = c("ErrDiffDensity", "ConfSubsetSize_Mbp"),
  labels = c("#Error Diff Per Mbp", "SubSetSize (Mbp)"))
fig_hg002_stratify_topdiff_h38 <- ggplot(dat.hg002.small.var.stratify.grc.topdiff.melt.h38, aes(x=value, y=reorder(Subset.Rename, reorder), fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) + scale_fill_manual(values = color_codes[1]) +
    facet_grid(~variable, scales = "free") +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + theme(legend.position = "none", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig_hg002_stratify_topdiff_h38

if (save){
  ggsave("figures/0309/small_variants-bwa_dv-stratify-topdiff-h37.pdf", fig_hg002_stratify_topdiff_h37, width=6, height=2.5)
  ggsave("figures/0309/small_variants-bwa_dv-stratify-topdiff-h38.pdf", fig_hg002_stratify_topdiff_h38, width=6, height=2.5)
}
```

```{r small_variant_calling_stratified_hg002}
# df_hg002_bwa_dv_stratify <- do.call(rbind, lapply(c("v0.5.1/hg002/stratify/grch38.extended.csv", "v0.5.1/hg002/stratify/grch37.extended.csv", "v0.5.1/hg002/stratify/chm13_to_grch38.extended.csv", "v0.5.1/hg002/stratify/chm13_to_grch37.extended.csv"), read_csv, map=map_caller_name))

df_hg002_bwa_dv_stratify <- df_hg002_bwa_dv_stratify %>% filter(Filter == "PASS", Subtype == "*", Subset %in% stratify_subsets)
df_hg002_bwa_dv_stratify$name = factor(df_hg002_bwa_dv_stratify$name, levels=levelnames)
df_hg002_bwa_dv_stratify$Subset.Rename <- factor(hash::values(map_strat_subsets, keys=df_hg002_bwa_dv_stratify$Subset), levels = strat_names)

df_hg002_bwa_dv_stratify <- df_hg002_bwa_dv_stratify %>% mutate(h38 = ifelse(name %in% c("GRCh38", "CHM13-to-GRCh38"), "GRCh38", "GRCh37"), pair = (name %in% c("GRCh38", "CHM13-to-GRCh38") * 1) + (Type == "SNP") * 10)

# fig_hg002_bwa_dv_stratify_scatter <- ggplot(df_hg002_bwa_dv_stratify, aes(y=METRIC.Recall, x=METRIC.Precision, color=name, shape=Type)) +
#   geom_path(aes(group = pair), color="#999999", arrow = arrow(length=unit(0.075, "inches"))) +
#   geom_point() +
#   facet_wrap(h38~Subset.Rename, ncol = 4, scales = "free_y", labeller = label_wrap_gen(multi_line=FALSE)) +
#   scale_color_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
#   ylab("Recall") + xlab("Precision") + labs(color="Method") + theme_bw() +
#   theme(axis.title.x=element_text(vjust=0, hjust=0.45),
#         axis.text.x = element_text(angle = 90, vjust = 1, hjust=0.8),
#         legend.position = "bottom")
# fig_hg002_bwa_dv_stratify_scatter

fig_hg002_bwa_dv_stratify_scatter_fixed <- ggplot(df_hg002_bwa_dv_stratify, aes(y=METRIC.Recall, x=METRIC.Precision, color=name, shape=Type)) + 
  geom_path(aes(group = pair), color="#999999", arrow = arrow(length=unit(0.075, "inches"))) +
  geom_point() + 
  facet_grid(~h38+Subset.Rename, labeller = label_wrap_gen(multi_line=FALSE)) +
  coord_cartesian(ylim=c(.84,1), xlim=c(0.84, 1)) +
  # facet_wrap(h38~Subset.Rename, ncol = 6, scales = "free_y", labeller = label_wrap_gen(multi_line=FALSE)) +
  # facet_grid(h38~Subset.Rename, scales = "free_y", labeller = label_wrap_gen(multi_line=FALSE)) +
  scale_color_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
  ylab("Recall") + xlab("Precision") + labs(color="Method") + theme_bw() +
  theme(axis.title.x=element_text(vjust=0, hjust=0.45),
        axis.text.x = element_text(angle = 90, vjust = 1, hjust=0.8),
        legend.position = "right", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig_hg002_bwa_dv_stratify_scatter_fixed

if (save){
  ggsave("figures/0309/small_variants-bwa_dv-stratify-mixed-scatter-facet-fixed.pdf", 
         fig_hg002_bwa_dv_stratify_scatter_fixed + theme(legend.position = "none"), width=8, height=4)
}

```

```{r small_variant_wgs_gatk_top_strata_hg002}
df.hg002.bwa.gatk.stratify.grc <- do.call(rbind, lapply(c("chm13v2-gatk/hg002/stratify/grch38.extended.csv", "chm13v2-gatk/hg002/stratify/hg19.extended.csv"), read_csv, map=map_caller_name)) %>% filter(Subtype == "*", Filter == "PASS", Subset.IS_CONF.Size > 10e6, Subset %in% strat_subsets)
df.hg002.bwa.gatk.stratify.leviosam <- do.call(rbind, lapply(c("chm13v2-gatk/hg002/stratify/chm13_to_grch38.extended.csv", "chm13v2-gatk/hg002/stratify/chm13_to_hg19.extended.csv"), read_csv, map=map_caller_name)) %>% filter(Subtype == "*", Filter == "PASS", Subset.IS_CONF.Size > 10e6, Subset %in% strat_subsets)

df.hg002.bwa.gatk.stratify.grc$Subset.Rename <- factor(hash::values(map_strat_subsets, keys=df.hg002.bwa.gatk.stratify.grc$Subset), levels = strat_names)
df.hg002.bwa.gatk.stratify.leviosam$Subset.Rename <- factor(hash::values(map_strat_subsets, keys=df.hg002.bwa.gatk.stratify.leviosam$Subset), levels = strat_names)
df.hg002.bwa.gatk.stratify.grc$ErrDiff <- df.hg002.bwa.gatk.stratify.grc$TRUTH.FN + df.hg002.bwa.gatk.stratify.grc$QUERY.FP - df.hg002.bwa.gatk.stratify.leviosam$TRUTH.FN - df.hg002.bwa.gatk.stratify.leviosam$QUERY.FP
df.hg002.bwa.gatk.stratify.leviosam$ErrDiff <- df.hg002.bwa.gatk.stratify.grc$TRUTH.FN + df.hg002.bwa.gatk.stratify.grc$QUERY.FP - df.hg002.bwa.gatk.stratify.leviosam$TRUTH.FN - df.hg002.bwa.gatk.stratify.leviosam$QUERY.FP
df.hg002.bwa.gatk.stratify.grc <- df.hg002.bwa.gatk.stratify.grc %>% mutate(ConfSubsetSize_Mbp = Subset.IS_CONF.Size / 1000000)
df.hg002.bwa.gatk.stratify.grc <- df.hg002.bwa.gatk.stratify.grc %>% mutate(ErrDiff, ErrDiffDensity = ErrDiff / ConfSubsetSize_Mbp)
df.hg002.bwa.gatk.stratify.leviosam <- df.hg002.bwa.gatk.stratify.leviosam %>% mutate(ConfSubsetSize_Mbp = Subset.IS_CONF.Size / 1000000)
df.hg002.bwa.gatk.stratify.leviosam <- df.hg002.bwa.gatk.stratify.leviosam %>% mutate(ErrDiff, ErrDiffDensity = ErrDiff / ConfSubsetSize_Mbp)

df.hg002.bwa.gatk.stratify <- rbind(df.hg002.bwa.gatk.stratify.grc, df.hg002.bwa.gatk.stratify.leviosam)



dat.hg002.bwa.gatk.stratify.grc.topdiff.h37 <- melt(rbind(df.hg002.bwa.gatk.stratify.grc %>% filter(name == "GRCh37", ErrDiffDensity > 1) %>% top_n(topdiff_n), df.hg002.bwa.gatk.stratify.grc %>% filter(name == "GRCh37", ErrDiffDensity < -1) %>% top_n(-topdiff_n)) %>% select(Type, Subtype, Subset.Rename, ConfSubsetSize_Mbp, ErrDiffDensity, name))
dat.hg002.bwa.gatk.stratify.grc.topdiff.h37$reorder <- rep(c(dat.hg002.bwa.gatk.stratify.grc.topdiff.h37 %>% filter(variable == "ErrDiffDensity"))$value, 2)
dat.hg002.bwa.gatk.stratify.grc.topdiff.h37$variable <- factor(
  dat.hg002.bwa.gatk.stratify.grc.topdiff.h37$variable,
  levels = c("ErrDiffDensity", "ConfSubsetSize_Mbp"),
  labels = c("#Error Diff Per Mbp", "SubSetSize (Mbp)"))
fig.hg002.bwa.gatk.stratify.grc.topdiff.h37 <- ggplot(dat.hg002.bwa.gatk.stratify.grc.topdiff.h37, aes(x=value, y=reorder(Subset.Rename, reorder), fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) + scale_fill_manual(values = color_codes[3]) +
    facet_grid(~variable, scales = "free") +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + theme(legend.position = "none", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.hg002.bwa.gatk.stratify.grc.topdiff.h37

dat.hg002.bwa.gatk.stratify.grc.topdiff.h38 <- melt(
  rbind(df.hg002.bwa.gatk.stratify.grc %>% filter(name == "GRCh38", ErrDiffDensity > 1) %>% top_n(topdiff_n), 
        df.hg002.bwa.gatk.stratify.grc %>% filter(name == "GRCh38", ErrDiffDensity < -1) %>% top_n(-topdiff_n)) %>% select(Type, Subtype, Subset.Rename, ConfSubsetSize_Mbp, ErrDiffDensity, name))
dat.hg002.bwa.gatk.stratify.grc.topdiff.h38$reorder <- rep(c(dat.hg002.bwa.gatk.stratify.grc.topdiff.h38 %>% filter(variable == "ErrDiffDensity"))$value, 2)
dat.hg002.bwa.gatk.stratify.grc.topdiff.h38$variable <- factor(
  dat.hg002.bwa.gatk.stratify.grc.topdiff.h38$variable,
  levels = c("ErrDiffDensity", "ConfSubsetSize_Mbp"),
  labels = c("#Error Diff Per Mbp", "SubSetSize (Mbp)"))
fig.hg002.bwa.gatk.stratify.grc.topdiff.h38 <- ggplot(dat.hg002.bwa.gatk.stratify.grc.topdiff.h38, aes(x=value, y=reorder(Subset.Rename, reorder), fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) + scale_fill_manual(values = color_codes[1]) +
    facet_grid(~variable, scales = "free") +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + theme(legend.position = "none", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.hg002.bwa.gatk.stratify.grc.topdiff.h38

if (save){
  ggsave("figures/0309/small_variants-bwa_gatk-stratify-topdiff-h37.pdf", fig.hg002.bwa.gatk.stratify.grc.topdiff.h37, width=6, height=2.5)
  ggsave("figures/0309/small_variants-bwa_gatk-stratify-topdiff-h38.pdf", fig.hg002.bwa.gatk.stratify.grc.topdiff.h38, width=6, height=2.5)
}
```

```{r small_variant_wgs_gatk_strata_hg002}
# df.hg002.bwa.gatk.stratify <- do.call(rbind, lapply(c("v0.5.1/hg002/stratify/grch38.extended.csv", "v0.5.1/hg002/stratify/grch37.extended.csv", "v0.5.1/hg002/stratify/chm13_to_grch38.extended.csv", "v0.5.1/hg002/stratify/chm13_to_grch37.extended.csv"), read_csv, map=map_caller_name))

df.hg002.bwa.gatk.stratify <- df.hg002.bwa.gatk.stratify %>% filter(Filter == "PASS", Subtype == "*", Subset %in% stratify_subsets)
df.hg002.bwa.gatk.stratify$name = factor(df.hg002.bwa.gatk.stratify$name, levels=levelnames)
df.hg002.bwa.gatk.stratify$Subset.Rename <- factor(hash::values(map_strat_subsets, keys=df.hg002.bwa.gatk.stratify$Subset), levels = strat_names)

df.hg002.bwa.gatk.stratify <- df.hg002.bwa.gatk.stratify %>% mutate(h38 = ifelse(name %in% c("GRCh38", "CHM13-to-GRCh38"), "GRCh38", "GRCh37"), pair = (name %in% c("GRCh38", "CHM13-to-GRCh38") * 1) + (Type == "SNP") * 10)


fig.hg002.bwa.gatk.stratify <- ggplot(df.hg002.bwa.gatk.stratify, aes(y=METRIC.Recall, x=METRIC.Precision, color=name, shape=Type)) + 
  geom_path(aes(group = pair), color="#999999", arrow = arrow(length=unit(0.075, "inches"))) +
  geom_point() + 
  facet_grid(~h38+Subset.Rename, labeller = label_wrap_gen(multi_line=FALSE)) +
  coord_cartesian(ylim=c(.7, 1), xlim=c(.7, 1)) +
  # facet_wrap(h38~Subset.Rename, ncol = 6, scales = "free_y", labeller = label_wrap_gen(multi_line=FALSE)) +
  # facet_grid(h38~Subset.Rename, scales = "free_y", labeller = label_wrap_gen(multi_line=FALSE)) +
  scale_color_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
  ylab("Recall") + xlab("Precision") + labs(color="Method") + theme_bw() +
  theme(axis.title.x=element_text(vjust=0, hjust=0.45),
        axis.text.x = element_text(angle = 90, vjust = 1, hjust=0.8),
        legend.position = "right", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.hg002.bwa.gatk.stratify

if (save){
  # ggsave("figures/small_variants-bwa_dv-stratify-mixed-bar.pdf", fig_hg002_bwa_dv_stratify_bars, width=10, height=8)
  ggsave("figures/0309/small_variants-bwa_gatk-stratify-mixed-scatter-facet-fixed.pdf", 
         fig.hg002.bwa.gatk.stratify + theme(legend.position = "none"), width=8, height=4)
}

```


# Pacbio

## Small variants -- all regions

```{r pacbio_small_variants}
df.hg002.hifi.dv <- do.call(rbind, lapply(c("pacbio/grch38.summary.csv", "pacbio/chm13_to_grch38.summary.csv", "pacbio/grch37.summary.csv", "pacbio/chm13_to_hg19.summary.csv"), read_csv, map=map_caller_name))
df.hg002.hifi.dv$name = factor(df.hg002.hifi.dv$name, levels=levelnames)

df.hg002.hifi.dv.melt <- melt(df.hg002.hifi.dv %>% select(name, Type, METRIC.Recall, METRIC.Precision, METRIC.F1_Score))#metric_orig))
df.hg002.hifi.dv.melt$variable <- factor(hash::values(map_metrics, keys=df.hg002.hifi.dv.melt$variable), levels = metric_names)
df.hg002.hifi.dv.melt$GrcSystem <- factor(hash::values(map_grc_system, keys=df.hg002.hifi.dv.melt$name))
fig.hg002.hifi.dv <- ggplot(df.hg002.hifi.dv.melt, aes(x=variable, y=value, fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) +
    scale_fill_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
    facet_grid(GrcSystem~Type) +
    # facet_grid(Type~.) +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + coord_cartesian(ylim=c(0.9,1)) +
    theme(axis.title.x=element_text(vjust=0, hjust=0.45), legend.position = "right", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.hg002.hifi.dv 

df.hg002.hifi.dv %>% filter(Filter == "PASS") %>% mutate(Err = TRUTH.FN + QUERY.FP) %>% select(name, Type, metric_orig, Err)
df.hg002.hifi.dv %>% filter(Filter == "PASS") %>% mutate(Err = TRUTH.FN + QUERY.FP) %>% select(name, Type, metric_orig, Err) %>% group_by(name) %>% summarise(Err=sum(Err))

if (save){
  ggsave("figures/0309/small_variants-pacbio.pdf", fig.hg002.hifi.dv+theme(legend.position = "none"), width=5, height=5)
  ggsave("figures/0309/small_variants-pacbio-2.pdf", fig.hg002.hifi.dv+theme(legend.position = "bottom"), width=8, height=5)
}
```

## Small variants -- CMRG regions

```{r pacbio_small_variants}
df.hg002.hifi.dv.cmrg <- do.call(rbind, lapply(c("pacbio/cmrg/grch38.summary.csv", "pacbio/cmrg/chm13_to_grch38.summary.csv", "pacbio/cmrg/grch37.summary.csv", "pacbio/cmrg/chm13_to_hg19.summary.csv"), read_csv, map=map_caller_name))
df.hg002.hifi.dv.cmrg$name = factor(df.hg002.hifi.dv.cmrg$name, levels=levelnames)

df.hg002.hifi.dv.cmrg.melt <- melt(df.hg002.hifi.dv.cmrg %>% select(name, Type, METRIC.Recall, METRIC.Precision, METRIC.F1_Score))#metric_orig))
df.hg002.hifi.dv.cmrg.melt$variable <- factor(hash::values(map_metrics, keys=df.hg002.hifi.dv.cmrg.melt$variable), levels = metric_names)
df.hg002.hifi.dv.cmrg.melt$GrcSystem <- factor(hash::values(map_grc_system, keys=df.hg002.hifi.dv.cmrg.melt$name))
fig.hg002.hifi.dv.cmrg <- ggplot(df.hg002.hifi.dv.cmrg.melt, aes(x=variable, y=value, fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) +
    scale_fill_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
    facet_grid(GrcSystem~Type) +
    # facet_grid(Type~.) +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + coord_cartesian(ylim=c(0.8,1)) +
    theme(axis.title.x=element_text(vjust=0, hjust=0.45), legend.position = "right", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.hg002.hifi.dv.cmrg 

df.hg002.hifi.dv.cmrg %>% filter(Filter == "PASS") %>% mutate(Err = TRUTH.FN + QUERY.FP) %>% select(name, Type, metric_orig, Err)
df.hg002.hifi.dv.cmrg %>% filter(Filter == "PASS") %>% mutate(Err = TRUTH.FN + QUERY.FP) %>% select(name, Type, metric_orig, Err) %>% group_by(name) %>% summarise(Err=sum(Err))

if (save){
  ggsave("figures/0309/small_variants-pacbio-cmrg.pdf", fig.hg002.hifi.dv.cmrg+theme(legend.position = "none"), width=5, height=5)
}
```

## Structural variants
```{r pacbio_struct_variants_h37}
df.hg002.hifi.sv.h37 <- do.call(rbind, lapply(c("pacbio/sv/grch37.summary.csv", "pacbio/sv/chm13_to_grch37.summary.csv"), read_csv, map=map_caller_name))
df.hg002.hifi.sv.h37$name = factor(df.hg002.hifi.sv.h37$name, levels=levelnames)

df.hg002.hifi.sv.h37.melt <- melt(df.hg002.hifi.sv.h37 %>% select(name, recall, precision, f1))#metric_orig))
df.hg002.hifi.sv.h37.melt$variable <- factor(hash::values(map_metrics, keys=df.hg002.hifi.sv.h37.melt$variable), levels = metric_sv_names)
df.hg002.hifi.sv.h37.melt$GrcSystem <- factor(hash::values(map_grc_system, keys=df.hg002.hifi.sv.h37.melt$name))
fig.hg002.hifi.sv.h37 <- ggplot(df.hg002.hifi.sv.h37.melt, aes(x=variable, y=value, fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) +
    scale_fill_manual(values = c("#E7B800", "#52854C")) +
    # facet_grid(GrcSystem~.) +
    # facet_grid(Type~.) +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + coord_cartesian(ylim=c(0.9,1)) +
    theme(axis.title.x=element_text(vjust=0, hjust=0.45), legend.position = "right", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.hg002.hifi.sv.h37 

df.hg002.hifi.sv.h37 %>% mutate(Err = FN + FP) %>% select(name, recall, precision, f1, Err)
df.hg002.hifi.sv.h37 %>% mutate(Err = FN + FP) %>% select(name, recall, precision, f1, Err) %>% group_by(name) %>% summarise(Err=sum(Err))

if (save){
  ggsave("figures/0309/SV-pacbio-h37.pdf", fig.hg002.hifi.sv.h37+theme(legend.position = "none"), width=3, height=5)
}
```

```{r pacbio_struct_variants_h38}
df.hg002.hifi.sv.h38 <- do.call(rbind, lapply(c("pacbio/sv/grch38.summary.csv", "pacbio/sv/chm13_to_grch38.summary.csv"), read_csv, map=map_caller_name))
df.hg002.hifi.sv.h38$name = factor(df.hg002.hifi.sv.h38$name, levels=levelnames)

df.hg002.hifi.sv.h38.melt <- melt(df.hg002.hifi.sv.h38 %>% select(name, recall, precision, f1))#metric_orig))
df.hg002.hifi.sv.h38.melt$variable <- factor(hash::values(map_metrics, keys=df.hg002.hifi.sv.h38.melt$variable), levels = metric_sv_names)
df.hg002.hifi.sv.h38.melt$GrcSystem <- factor(hash::values(map_grc_system, keys=df.hg002.hifi.sv.h38.melt$name))
fig.hg002.hifi.sv.h38 <- ggplot(df.hg002.hifi.sv.h38.melt, aes(x=variable, y=value, fill=name)) + 
    geom_bar(stat="identity", position=position_dodge()) +
    scale_fill_manual(values = c("#FC4E07", "#00AFBB", "#E7B800", "#52854C")) +
    ylab("") + xlab("") + labs(fill="Method") + theme_bw() + coord_cartesian(ylim=c(0.9,1)) +
    theme(axis.title.x=element_text(vjust=0, hjust=0.45), legend.position = "right", strip.background = element_rect(color="white", fill="white", size=0.5, linetype="solid"), strip.text = element_text(face="bold"))
fig.hg002.hifi.sv.h38 

df.hg002.hifi.sv.h38 %>% mutate(Err = FN + FP) %>% select(name, recall, precision, f1, Err)
df.hg002.hifi.sv.h38 %>% mutate(Err = FN + FP) %>% select(name, recall, precision, f1, Err) %>% group_by(name) %>% summarise(Err=sum(Err))

if (save){
  ggsave("figures/0309/SV-pacbio-h38.pdf", fig.hg002.hifi.sv.h38+theme(legend.position = "none"), width=3, height=5)
}
```

